specreduce/tracing.py

```diff
@@ -37,6 +38,11 @@ def __post_init__(self):
     def __getitem__(self, i):
         return self.trace[i]
 
+    def _parse_image(self):
+        if isinstance(self.image, Spectrum1D):
+            # NOTE: should the Spectrum1D structure instead be preserved?
+            self.image = self.image.data
```

Comment:
This drops the unit. Are you thinking of saving it as a class attribute or something else?

Reply:
Well this is the main reason why I was wondering if we should keep the quantity/NDData/spectrum1D internally as opposed to the separate unit parameter, that way we can keep the unit internally and avoid all the parsing logic.

Reply:
That will be more involved, but it would probably also fix the Quantity issue I mentioned in my second standalone comment in the PR thread. I'll give it a try and see how it comes out.

Reply:
I think skipping the parsing logic pushes the problem farther down the road. `self.image` is just the flux when you have a CCDData/NDData/numpy array image but is an object containing the flux, spectral axis, and uncertainty when `self.image` is type Spectrum1D. The only way to put them all on the same footing is to take `image.data` of the NDData-inheriting types (which include Spectrum1D), which produces a unit-less array.

If you don't pre-process, then you have to check type any time you want to do an operation (e.g., if it's an array, I can take `image.mean()`, but if it's Spectrum1D, I have to do `image.flux.mean()` or `image.data.mean()` instead, etc.).

Reply:
ok, I agree then, some sort of parsing and coercing into a consistent internal type is probably the best way to go.  The questions that remain then are:
* what is the most convenient type to use internally?  If we ever are going to act on uncertainties, if provided, then maybe converting any input into `Spectrum1D` objects and storing that internally makes sense?
* do we always output the same type?  Or do we attempt to convert back to the type the user provided?  This might be out of scope for this PR, but at least worth planning ahead.
* does it make sense to still have the `unit` input as a dedicated parameter or require the user to pass an object that contains units if they don't want default units?  I think I prefer removing (probably with deprecation) the `unit` input parameter to avoid user-confusion and internal logic needed to handle the case where the parameter and a passed `Spectrum1D` or quantity object have conflicting units.

Reply:
To this point, the practice has been to give back `self.image` the same way the user provided it as its class' `image` argument. Then, inside the class, we use `self.image` to create a local variable that, if it's not already one, converts the image to a numpy array (either normal or masked) for internal operations.

If we accept Spectrum1D images, we probably need to start coercing all images to a consistent type before assigning them to `self.image`, since, as I mentioned above, calling a Spectrum1D object returns more than just its flux. An upside of this approach is that it would also allow us to use `self.image` for internal operations instead of a converted local variable since we'd now be sure of its type. 

CCDData is my pick as the best type for that purpose. Working internally with unit-less numpy arrays would be simpler, but if a user provides an image with units, they should be able to see those units when they call `self.image`. Unlike Quantity arrays, you could also get the uncertainty if you dig deeper (`self.image.uncertainty`), which is a plus.

`HorneExtract` is the only class that takes a `unit` argument, and it was requested so that authors had the ability to pass a string unit with a regular numpy array and get back a spectrum with Astropy units. The `unit` argument only matters if `image` is a numpy array; like `variance` and `mask`, it is skipped otherwise.

Reply:
CCDData sounds like a reasonable pick to me.  The other option that might be worth considering is Spectrum1D since I think some methods currently always return a Spectrum1D and these packages are intended to coordinate well with each other.  If that comes with more overhead though, then using CCDData internally and mapping to Spectrum1D when necessary for output works for me.

> HorneExtract is the only class that takes a unit argument, and it was requested so that authors had the ability to pass a string unit with a regular numpy array and get back a spectrum with Astropy units. The unit argument only matters if image is a numpy array; like variance and mask, it is skipped otherwise.

I would still be in favor of deprecating this, but it doesn't have to happen in this PR.  If we ever did get rid of it, the user would still be able to use `numpy_array * u.Unit(unit_string)`, and I think it would considerably clean up internal logic once we were able to get rid of it fully.  Alternatively, adding matching support across all other classes would either require duplicate code logic or refactoring with some base classes or decorators that handle that in general.